### PR TITLE
Feature/mspcg deflation

### PR DESCRIPTION
**Pull request by cpviolator**
This PR adds a new Dirac operator to QUDA, matEig. Those solvers with the capability to deflate will now use the operator constructed by `gauge(Long/Fat)Eigensolver (and clover field too). New command lines have been added:
```
 --prec-eigensolver ENUM:value in {double,single,half,quarter}=-2147483648
 --recon-eigensolver ENUM:value in {18,13,12,9,8}=-2147483648
```
to dictate the precision of the eigensolver when used as a deflation operator. Two command line commands have been tested:
```
mpirun -np 2 ./invert_test --sdim 16 --tdim 32 --tgridsize 2 --inv-deflate true --load-gauge /scratch/lattices/wl_16_64_5p5_x2p38_um0p4086_cfg_1000.lime --mass -0.42 --verbosity verbose --tol 1e-10 --inv-type gcr --anisotropy 2.38 --solve-type direct-pc --prec double --prec-sloppy half --prec-precondition half --prec-eigensolver single --ngcrkrylov 50 --niter 5000 --verify false --eig-tol 1e-3 --eig-poly-deg 50 --eig-amin 1e-3 --eig-n-ev 32 --eig-n-kr 64
```
and
```
mpirun -np 2 ./invert_test --sdim 16 --tdim 32 --tgridsize 2 --inv-deflate true --load-gauge /scratch/lattices/wl_16_64_5p5_x2p38_um0p4086_cfg_1000.lime --mass -0.42 --verbosity verbose --tol 1e-10 --inv-type cg --anisotropy 2.38 --solve-type normop-pc --prec double --prec-sloppy half --prec-precondition half --prec-eigensolver single --ngcrkrylov 50 --niter 5000 --verify false --eig-tol 1e-3 --eig-poly-deg 50 --eig-amin 1e-3 --eig-n-ev 32 --eig-n-kr 64
```
which test the SVD and regular deflation code paths.

@walkloud This will now allow us to deflate a solve using the MSPCG inverter.

### CODE REVIEW

**Comment by maddyscientist** [comment]
Great addition, thanks!  @cpviolator have you tested this is working ok in combination with MSPCG?  Perhaps you could run the same tests that @walkloud was running in #1027?

**Comment by cpviolator** [comment]
Just picked up an error I'm not familiar with:
```
Tuned block=(128,1,1), grid=(80,6,1), shared_bytes=0, aux=(1,0,0,0) giving 715.03 Gflop/s, 1430.07 GB/s for 16N4quda4blas9multiCdotIdfEE with policy_kernel,vol=393216,stride=393\
216,precision=4,order=4,Ns=4,Nc=3
ERROR: Compile-time 252 and run-time 170 limits disagree (rank 0, host a02n13, /ccs/home/howarth/QUDA_MSPCG_DEFL/quda/lib/../include/multi_blas_helper.cuh:168 in staticCheck())
       last kernel called was (name=16N4quda4blas9multiCdotIdfEE,volume=32x16x8x8x12,aux=policy_kernel,vol=393216,stride=393216,precision=4,order=4,Ns=4,Nc=3)
```

I've attached the `CMakeCache.txt` too
[CMakeCache.txt](https://github.com/lattice/quda/files/4905553/CMakeCache.txt)

This occurred in the eigensolver with the following command line args:
```
--dslash-type mobius --dim 64 16 8 8 --gridsize 1 4 8 12 --nsrc 5 --Lsdim 12 --m5 -1.1 --b5 1.5 --c5 0.5 --mass 0.00152 --load-gauge /gpfs/alpine/proj-shared/lgt100/c51/x_files/project_2/production/a09m135_a/cfgs_flow/l6496f211b630m001326m03636m4313a.78_wflow1.0.lime --verbosity verbose --solution-type mat --matpc even-even --solve-type normop-pc --prec double --recon 12 --niter 14000 --prec-sloppy half --reliable-delta 0.1 --prec-precondition half --prec-eigensolver single --recon-eigensolver 12 --inv-type pcg --precon-type cg --precon-schwarz-type additive --tol-precondition 1e-6 --maxiter-precondition 5 --inv-deflate true --eig-amin 2.0e-5 --eig-poly-deg 400 --eig-n-ev 300 --eig-n-conv 300 --eig-n-kr 350 --eig-tol 1e-6 --eig-batched-rotate 5 --eig-max-restarts 100 --eig-require-convergence true --eig-use-poly-acc true --eig-spectrum SR --eig-use-normop true --eig-save-vec /gpfs/alpine/scratch/howarth/lgt104/test_file_300
```

**Comment by maddyscientist** [comment]
Can you merge in the latest develop, this should fix this.

**Comment by cpviolator** [comment]
Thank you!

File: lib/multigrid.cpp
Comment (maddyscientist) [inline lib/multigrid.cpp]: delete comment

File: lib/inv_gmresdr_quda.cpp
Comment (maddyscientist) [inline lib/inv_gmresdr_quda.cpp:25]: A lot of reformatting of this file.  Any reasoning behind this?  Since @alexstrel  is working on cleaning up his deflation algorithms, this will likely lead to a lot of conflicts for him.

File: include/dirac_quda.h
Comment (maddyscientist) [inline include/dirac_quda.h:1893]: Is this comment correct, that deflation will use `dRef`?

File: lib/inv_gmresdr_quda.cpp
Comment (cpviolator) [reply lib/inv_gmresdr_quda.cpp:25]: I'm pretty sure that's just a clang format.

Comment (maddyscientist) [reply lib/inv_gmresdr_quda.cpp:25]: Sure, but clang-format should only be passed to the diff from develop.  This looks to be touching lots of untouched lines.

**Review by maddyscientist — COMMENTED** [review]
I've done my initial code review.  Looks to be in good shape as is.  Beyond inline comments, one thing comes to mind: can we get rid of the "ritz" precision, and just use the  "eigensolver" precision wherever the Ritz precision is currently used?

Will doing some run tests next.

Thanks again for this contribution @cpviolator!

File: lib/inv_gmresdr_quda.cpp
Comment (cpviolator) [reply lib/inv_gmresdr_quda.cpp:25]: Looks like I haven't 'clanged' this yet. I'll apply the formatting script, should bring it in line with develop.

Comment (cpviolator) [reply lib/inv_gmresdr_quda.cpp:25]: Hmm. I guess not re clang. I did have to touch this file to rename eigensolver variables. Not sure why so many lines got changed.

**Comment by cpviolator** [comment]
> I've done my initial code review. Looks to be in good shape as is. Beyond inline comments, one thing comes to mind: can we get rid of the "ritz" precision, and just use the "eigensolver" precision wherever the Ritz precision is currently used?
> 
> Will doing some run tests next.
> 
> Thanks again for this contribution @cpviolator!

@alexstrel Do you foresee any issues in replacing the `cuda_prec_ritz` variable with `eigensolver_prec`? Other than parameter setting and header files, the only place it exists is `deflation.cpp`

**Comment by cpviolator** [comment]
@maddyscientist @alexstrel So it looks like the majority of instances of `cuda_prec_ritz` are in the `eigCG` and related code paths. If you're happy to delay this change until the merge of the `eigCG` code I'd be happy to deal with it on that PR.

**Comment by maddyscientist** [comment]
Yep, I'm happy deferring removal of `cuda_prec_ritz` until Alexei's PR.

@cpviolator note that you will need to merge in develop to get your built tests working again: Eigen download broke due to Eigen switching from bitbucket to gitlab.  If you merge in develop, all the build tests will work again.

One other thing I just noticed: this PR will break the Fortran interface.  Make sure that all additions to quda.h and enum_quda.h are matched with quda_fortran.F90 and enum_quda_fortran.h, respectively.

**Review by maddyscientist — APPROVED** [review]
I've done a good amount of testing with this, and it's all looking good.  In particular I've done some testing of a quad-precision deflated preconditioned CG solver using Lanczos and MSPCG combined 😄.  This PR make that possible, and all is looking good.  

Just running some small lattice (16^3x64x8) tests on nvsocal2 with 2 GPUs, I see the following:
*regular CG* (double-half)
```12 solves, mean iteration count 3694.55 (stddev = 53.4254), with mean solve time 12.3805 (stddev = 0.180583), mean GFLOPS 3428.16 (stddev = 10.5061) [excluding first solve]```

*deflated CG* (double-half, single precision e-vectors)
```12 solves, mean iteration count 617.182 (stddev = 1.60114), with mean solve time 2.3613 (stddev = 0.00572853), mean GFLOPS 3130.27 (stddev = 2.22064) [excluding first solve]```

*MSPCG* (double-half-quarter)
```12 solves, mean iteration count 942.455 (stddev = 8.89228), with mean solve time 13.3386 (stddev = 0.119939), mean GFLOPS 10307 (stddev = 32.4175) [excluding first solve]```

*deflated MSPCG* (double-half-quarter, single precision e-vectors)
```12 solves, mean iteration count 163.182 (stddev = 0.750757), with mean solve time 2.57983 (stddev = 0.0110491), mean GFLOPS 9385.42 (stddev = 17.0759) [excluding first solve]```

Note that the combined effect of MSPCG and deflation is nicely multiplicative as we would desire.

@hummingtree, @mathiaswagner or @weinbe2 do you have anything to add here?

**Comment by hummingtree** [comment]
I am a little surprised that MSPCG almost breaks even on 2 GPUs?
~~And the preconditioner works better in terms of reducing iteration count (`3694/942=3.9 < 617/163=3.8`) when deflation is in place. This aligns with the expectation that "MSPCG preconditioner takes care of the high modes and deflation takes care of the low modes".~~
Thanks @cpviolator - this addition could be very useful.

**Comment by maddyscientist** [comment]
I was using GPUs 1 and 2 on nvsocal2.  P2P bandwidth is probably a good Summit simulator 😉 
```
*************************************************************************
Test Description: Bus bandwidth between pairs of devices
*************************************************************************
P2P Bandwidth Matrix (GB/s) - Unidirectional, P2P=Enabled
   D\D      0      1      2
      0 265.16  10.39   6.78
      1  10.39 265.00   6.78
      2   5.21   5.21 265.31
P2P Bandwidth Matrix (GB/s) - Bidirectional, P2P=Enabled
   D\D      0      1      2
      0 279.22  20.31  10.42
      1  20.31 277.04   5.96
      2  10.40   5.96 276.84
```

**Comment by hummingtree** [comment]
But you see it's still better than Summit considering there are `6` Volta's sharing one `23GB/s` NIC.

File: lib/check_params.h
Comment (weinbe2) [inline lib/check_params.h:518]: Not here specifically, but in this file, `cuda_prec_eigensolver` needs to be appropriately init/checked.

Comment (weinbe2) [reply lib/check_params.h:518]: This is across `newQudaInvertParam`, `newQudaGaugeParam`, and `newQudaCloverParam`

(Sticking a pin in the thought that we should see if there's a way we can automate checking if each field in the structs have an appropriate entry in these functions.)

File: tests/staggered_invert_test.cpp
Comment (weinbe2) [inline tests/staggered_invert_test.cpp]: `iter[k]`?

File: lib/check_params.h
Comment (weinbe2) [reply lib/check_params.h:518]: Ah, also missed that it's needed for `reconstruct_eigensolver` as well.

**Comment by weinbe2** [comment]
Has this been tested with staggered (naive and asqtad, latter to cover fat/long fields) and with the clover op (to cover clover fields)?

**Comment by cpviolator** [comment]
@weinbe2 I tested it with wilson, clover, twisted, twisted clover, and mobius. I'll spin up with staggered tonight and run a few jobs through it.

**Comment by weinbe2** [comment]
> @weinbe2 I tested it with wilson, clover, twisted, twisted clover, and mobius. I'll spin up with staggered tonight and run a few jobs through it.

That'd be great @cpviolator , thank you.

**Comment by weinbe2** [comment]
Has this gone through a `SANITIZE` build? @maddyscientist

**Comment by cpviolator** [comment]
Picking up a link error when I try to build this on NVSOCAL with staggered:
```
cpviolator@nvsocal2:/scratch/CPviolator/work/QUDA_MSPCG_DEFL/build$ make
[  3%] Built target Eigen
[ 28%] Built target quda_cpp
[ 28%] Built target quda_generic_target
[ 30%] Built target quda_cuda_target
[ 30%] Linking CXX shared library libquda.so
	/usr/local/cuda/lib64/libcudart_static.a(libcudart_static.a.o): in function `cudaguidstructstampinit()':
cuda_runtime_api.cpp:(.text.startup+0x13): relocation truncated to fit: R_X86_64_PC32 against `.data'
cuda_runtime_api.cpp:(.text.startup+0x1a): relocation truncated to fit: R_X86_64_PC32 against `.data'
/usr/local/cuda/lib64/libcudart_static.a(libcudart_static.a.o): in function `_GLOBAL__sub_I.00150_cudart_global.cpp':
cudart_global.cpp:(.text.startup+0x7a): relocation truncated to fit: R_X86_64_PC32 against `.bss'
cudart_global.cpp:(.text.startup+0x91): relocation truncated to fit: R_X86_64_PC32 against `.bss'
/usr/lib/gcc/x86_64-linux-gnu/9/crtbeginS.o: in function `deregister_tm_clones':
crtstuff.c:(.text+0x3): relocation truncated to fit: R_X86_64_PC32 against `.tm_clone_table'
crtstuff.c:(.text+0xa): relocation truncated to fit: R_X86_64_PC32 against symbol `__TMC_END__' defined in .nvFatBinSegment section in libquda.so
/usr/lib/gcc/x86_64-linux-gnu/9/crtbeginS.o: in function `register_tm_clones':
crtstuff.c:(.text+0x33): relocation truncated to fit: R_X86_64_PC32 against `.tm_clone_table'
crtstuff.c:(.text+0x3a): relocation truncated to fit: R_X86_64_PC32 against symbol `__TMC_END__' defined in .nvFatBinSegment section in libquda.so
/usr/lib/gcc/x86_64-linux-gnu/9/crtbeginS.o: in function `__do_global_dtors_aux':
crtstuff.c:(.text+0x72): relocation truncated to fit: R_X86_64_PC32 against `.bss'
crtstuff.c:(.text+0x9a): relocation truncated to fit: R_X86_64_PC32 against `.bss'
CMakeFiles/quda.dir/dslash_coarse.cu.o: in function `__nv_save_fatbinhandle_for_managed_rt(void**)':
/scratch/CPviolator/work/QUDA_MSPCG_DEFL/quda/lib/dslash_coarse.cu:1:(.text+0xf): additional relocation overflows omitted from the output
collect2: error: ld returned 1 exit status
make[2]: *** [lib/CMakeFiles/quda.dir/build.make:1937: lib/libquda.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:1118: lib/CMakeFiles/quda.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
cpviolator@nvsocal2:/scratch/CPviolator/work/QUDA_MSPCG_DEFL/build$ 
```

Here is the  
[CMakeCache.txt](https://github.com/lattice/quda/files/5142779/CMakeCache.txt)

I'll try on Summit and/or Lassen.

**Comment by cpviolator** [comment]
OK, ran this through the sanitizer and used the following commands to test staggered:
```
./staggered_invert_test --dim 16 16 16 16 --gridsize 1 1 1 1 --verbosity verbose --nsrc 3 --dslash-type asqtad --compute-fat-long true --mass 0.2 --tadpole-coeff 0.9 --tol 1e-6 --prec double --niter 10000 --prec-eigensolver double --eig-use-poly-acc TRUE --eig-use-normop true --eig-max-restarts 100000 --eig-tol 1e-6 --eig-amin 1.601e-1 --eig-spectrum SR --eig-poly-deg 200 --load-gauge /scratch/weinbe2/configs_b7p0/l16t16b7p0 --eig-n-kr 64 --eig-n-ev 16 --recon 18 --prec-sloppy single --recon-sloppy 13 --test 1 --inv-deflate true
```

```
./staggered_eigensolve_test --dim 16 16 16 16 --gridsize 1 1 1 1 --verbosity verbose --nsrc 3 --dslash-type asqtad --compute-fat-long true --mass 0.2 --tadpole-coeff 0.9 --tol 1e-6 --prec double --niter 10000 --prec-eigensolver double --eig-use-poly-acc false --eig-use-normop false --eig-max-restarts 100000 --eig-tol 1e-6 --eig-amin 1.601e-1 --eig-spectrum SR --load-gauge /scratch/weinbe2/configs_b7p0/l16t16b7p0 --eig-n-kr 64 --eig-n-ev 16 --recon 18 --prec-sloppy single --recon-sloppy 13 --test 3
```

All seems well on both fronts.

**Review by weinbe2 — APPROVED** [review]
Thanks @cpviolator , looks good. This will be good to merge once Jenkins completes.

**Pull request merged**
